core/src/main/java/org/infinispan/util/concurrent/locks/impl/DefaultPendingLockManager.java

```diff
@@ -0,0 +1,373 @@
+
+import org.infinispan.commons.util.Notifier;
+import org.infinispan.context.impl.TxInvocationContext;
+import org.infinispan.factories.annotations.ComponentName;
+import org.infinispan.factories.annotations.Inject;
+import org.infinispan.transaction.impl.LocalTransaction;
+import org.infinispan.transaction.impl.TransactionTable;
+import org.infinispan.transaction.xa.CacheTransaction;
+import org.infinispan.transaction.xa.GlobalTransaction;
+import org.infinispan.util.TimeService;
+import org.infinispan.util.concurrent.TimeoutException;
+import org.infinispan.util.concurrent.locks.PendingLockListener;
+import org.infinispan.util.concurrent.locks.PendingLockManager;
+import org.infinispan.util.concurrent.locks.PendingLockPromise;
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Predicate;
+import java.util.stream.Collector;
+import java.util.stream.Collectors;
+
+import static java.lang.String.format;
+import static org.infinispan.commons.util.InfinispanCollections.findAnyCommon;
+import static org.infinispan.commons.util.Util.toStr;
+import static org.infinispan.factories.KnownComponentNames.TIMEOUT_SCHEDULE_EXECUTOR;
+
+/**
+ * The default implementation for {@link PendingLockManager}.
+ * <p>
+ * In transactional caches, a transaction would wait for transaction originated in a older topology id. It can happen
+ * when topology changes and a backup owner becomes the primary owner.
+ *
+ * @author Pedro Ruivo
+ * @since 8.0
+ */
+public class DefaultPendingLockManager implements PendingLockManager {
+
+   private static final Log log = LogFactory.getLog(DefaultPendingLockManager.class);
+   private static final boolean trace = log.isTraceEnabled();
+   private static final int NO_PENDING_CHECK = -2;
+   private final Map<GlobalTransaction, PendingLockPromiseImpl> pendingLockPromiseMap;
+   private TransactionTable transactionTable;
+   private TimeService timeService;
+   private ScheduledExecutorService timeoutExecutor;
+
+   public DefaultPendingLockManager() {
+      pendingLockPromiseMap = new ConcurrentHashMap<>();
+   }
+
+   @Inject
+   public void inject(TransactionTable transactionTable, TimeService timeService,
+                      @ComponentName(TIMEOUT_SCHEDULE_EXECUTOR) ScheduledExecutorService timeoutExecutor) {
+      this.transactionTable = transactionTable;
+      this.timeService = timeService;
+      this.timeoutExecutor = timeoutExecutor;
+   }
+
+   @Override
+   public PendingLockPromise checkPendingForKey(TxInvocationContext<?> ctx, Object key, long time, TimeUnit unit) {
+      final GlobalTransaction globalTransaction = ctx.getGlobalTransaction();
```

Comment:
A general question for this class, but it seems the purpose is just to wait for locks to be released not to actually acquire them.  Normally when I unblock like this I would expect the lock to be acquired as well.  But it seems this is still up to the LockManager to perform this duty?